### DISCUSSION TITLE
Question about RequireAdminOrRecordAuth (v0.8)

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#952] [Q&A]
https://github.com/pocketbase/pocketbase/blob/rc/apis/middlewares.go#L152-L169

> // You can further filter the allowed auth record collections by providing their names.
> //
> // This middleware is the opposite of [apis.RequireGuestOnly()].
> func RequireAdminOrRecordAuth(optCollectionNames ...string) echo.MiddlewareFunc {

What makes a record belonging to an auth record?  Is this defined by a relationship from one collection to another?  There aren't any examples for how Extending Pocketbase works using Golang code.  It's all Javascript API.  So I'm not sure what is going on here.

**Comment by ganigeorgiev** [comment]
> What makes a record belonging to an auth record?

That's up to the developers and it depends on your collections structure. For example, you can add a relation field `author` and point it to the `users` collection. And then in your Collection API rules, you could define a rule like: `author = @request.auth.id`.

The `RequireAdminOrRecordAuth` middleware doesn't express "ownership". It only requires the request to have a valid admin or auth record `Authorization` token. The request will be further filtered via the related collection API rule (if the accessed route is for some of the crud endpoints).

I'm currently working on adapting the documentation for v0.8.0 and there will be examples for common tasks when extending PocketBase (eg. operating with the database, defining custom routes, intercepting create/update, applying custom validators, etc.).

**Comment by jayhale** [comment]
@ganigeorgiev could you describe what is meant by the different auth header types (record, owner)?

I'm looking through the middleware docs and am a bit confused.  I'm currently working on securing a custom API route by requiring that the currently authorized user be affiliated with the record through `@request.auth.id = user.id`.

E.g., I want an endpoint like `/api/endpoints/checkcalendar/{calendarId}` where `Calendars` is a collection that has a `user` relation. The endpoint should return `Unauthorized` if the calendar record's user relation does not match the authorized user.

I'm not sure how to do this with the provided middleware.  It seems like this would be `apis.RequireAdminOrOwnerAuth`, but how to I indicate to compare the related user ID to the authorized user?

**Reply by ganigeorgiev** [reply]
The `apis.RequireAdminOrOwnerAuth(ownerIdParam)` middleware allows admins OR auth record that has id matching the specified path parameter.

This middleware can't be used in your case because the path parameter would be the `calendarId`, not the user id. Additionally note that the auth middlewares execute **before** the action handler, or in other words, we don't know whether the `calendarId` record exist or that it has a `user` relation field.

So in your case you'll have to define your own custom "post" action middleware (aka. process first the `next()` call and then do your custom checks) OR just handle the auth check in the endpoint (_I personally would go with this since it is simpler_), something like:

```go
package main

import (
	"log"
	"net/http"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/models"
)

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.AddRoute(echo.Route{
			Method: http.MethodGet,
			Path:   "/api/endpoints/checkcalendar/:calendarId",
			Handler: func(c echo.Context) error {
				calendar, err := app.Dao().FindRecordById("calendars", c.PathParam("calendarId"))
				if err != nil {
					return apis.NewNotFoundError("The calendar does not exist.", err)
				}

				// Check the auth record id with the calendar's user id.
				//
				// If authRecord is nil, then this mean that the client is authenticated as admin.
				// This is guaranteed because of the RequireAdminOrRecordAuth middleware.
				authRecord, _ := c.Get(apis.ContextAuthRecordKey).(*models.Record)
				if authRecord != nil && calendar.GetString("user") != authRecord.Id {
					// or NewNotFoundError if you don't want to show the user that this calendar exist to avoid enumerations
					return apis.NewForbiddenError("You are not allowed to access this calendar.", nil)
				}
				
				// do something else...

				// Enable ?expand query param support.
				apis.EnrichRecord(c, app.Dao(), calendar)

				return c.JSON(http.StatusOK, calendar)
			},
			Middlewares: []echo.MiddlewareFunc{
				apis.RequireAdminOrRecordAuth(),
			},
		})

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

**Reply by jayhale** [reply]
@ganigeorgiev tyty!
